tools/translate_annotations.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import re
+import subprocess
+from bisect import bisect_right
+from collections import defaultdict
+from typing import (Callable, DefaultDict, Generic, List, Optional, Pattern,
+                    Sequence, TypeVar, cast)
+
+from typing_extensions import TypedDict
+
+
+class Hunk(TypedDict):
+    old_start: int
+    old_count: int
+    new_start: int
+    new_count: int
+
+
+class Diff(TypedDict):
+    old_filename: str
+    hunks: List[Hunk]
+
+
+# adapted from the similar regex in tools/clang_tidy.py
+# @@ -start,count +start,count @@
+hunk_pattern = r'^@@\s+-(\d+)(?:,(\d+))?\s+\+(\d+)(?:,(\d+))?\s+@@'
+
+
+def parse_diff(diff: str) -> Diff:
+    name = None
+    hunks: List[Hunk] = []
+    for line in diff.splitlines():
+        name_match = re.match(r'^--- a/(.*)$', line)
+        hunk_match = re.match(hunk_pattern, line)
+        if name:
+            assert not name_match
+            if hunk_match:
+                old_start, old_count, new_start, new_count = hunk_match.groups()
+                hunks.append({
+                    'old_start': int(old_start),
+                    'old_count': int(old_count or '1'),
+                    'new_start': int(new_start),
+                    'new_count': int(new_count or '1'),
+                })
+        else:
+            assert not hunk_match
+            if name_match:
+                name, = name_match.groups()
+    assert name
+    return {
+        'old_filename': name,
+        'hunks': hunks,
+    }
+
+
+T = TypeVar('T')
+U = TypeVar('U')
+
+
+# https://gist.github.com/ericremoreynolds/2d80300dabc70eebc790
```

Comment:
Reference to gist is good, but quick inline explanation what it's for would be nice as well